util/codec/codec.go

```diff
@@ -714,6 +714,28 @@ func DecodeRange(b []byte, size int) ([]types.Datum, []byte, error) {
 	return values, nil, nil
 }
 
+// DecodeOneWithTp decodes on datum from a byte slice generated with EncodeKey or EncodeValue.
+func DecodeOneWithTp(b []byte, ft *types.FieldType) (remain []byte, d types.Datum, err error) {
+	remain, d, err = DecodeOne(b)
+	if ft.Tp != mysql.TypeNewDecimal || err != nil {
+		return remain, d, err
+	}
+	if ft.Decimal != types.UnspecifiedLength && d.Frac() != ft.Decimal && d.GetInterface() != nil {
```

Comment:
Why not do it in `DecodeOne`?

Reply:
Because `DecodeOne` doesn't know the `ft` parameter.